Add support for multi-GCS operation

Davidsastresas opened the pull request:
This PR adds support for the recently added multi-GCS mavlink subprotocol https://github.com/mavlink/mavlink/pull/2158. This allows graceful change in control ownership on a system with multiple GCS. Please note this subprotocol does not attempt to cover security, it assumes all the operators are in contact between them and they work in a collaborative manner.

For more information about the protocol itself, please read that thread instead. Mavlink docs will be updated shortly, when we confirm all of this looks good.

On this PR, the implementation is as follows:
### New top toolbar indicator
This indicator will be populated if the active vehicle is sending the new CONTROL_STATUS message. On this icon we can grasp:
- System Id in control
- If this GCS is currently in control
- If takeover is allowed automatically, or we need to ask to the GCS in control first.

Note this icon will have an animation effect whenever the control status changes, whether it is because of a change in takeover allowed, or because of a change in system in control.

This way, we have some variants:


![Screenshot from 2025-02-08 16-11-27](https://github.com/user-attachments/assets/6166a457-208a-4af9-a662-9144dd3d5416)
On this case this GCS (252) is in control, as we see the the sysId label green, and also the aircraft icon and the segment joining label and aircraft.

![Screenshot from 2025-02-08 16-16-03](https://github.com/user-attachments/assets/be14a281-274c-43da-9303-5502589ab29c)
On this case, this GCS (252) is not in control ( label is white, not green ), and also the aircraft icon and segment joining is white, indicating automatic takeover is not allowed ( we need to ask to the GCS in control first )

![Screenshot from 2025-02-08 16-18-01](https://github.com/user-attachments/assets/77d5bf2b-4a26-4de6-919e-b8ae27e23ad2)
On this case it is similar to the above situation, only automatic takeover is allowed ( aircraft icon in green ) so we don't need to request permission to the GCS in control, we can adquire control directly.

### Expanded menu after clicking top toolbar indicator
After clicking the top toolbar icon, we see an expanded panel. This panel changes depending on the particular control situation. On this panel, we can:
- See what system id is in control, and see it highlighted in green when it is us
- See if automatic takeover is allowed, or the requesting GCS needs to ask permision first
- Send control requests, in case we are not in control
- If we are in control, we can change if automatic takeover is allowed or not
- We can also change SystemID of this GCS from here, similar to how we usually do in the general settings -> telemetry settings
![Screenshot from 2025-02-08 16-19-49](https://github.com/user-attachments/assets/c672116c-5458-48eb-abb2-584a1d831f41)

## Control request procedure when takeover is allowed
In the most simple case where we are not in control and takeover is allowed, just clicking "Acquire control" will grant us control of the vehicle.
Note that on this situation, we can choose before acquiring control if we want to allow automatic takeover or not, on the tickbox. If we want to change this after being in control it is possible too, see previous screenshot. 
![Screenshot from 2025-02-08 16-24-19](https://github.com/user-attachments/assets/894c018c-32c4-4d12-b48b-223cc79ed333)

## Control request procedure when takeover is not allowed
On this situation, we will have the following panel:
![Screenshot from 2025-02-08 16-26-44](https://github.com/user-attachments/assets/c81d3f41-19e6-4593-9e47-e501134932cf)
Clicking "send request" will send a control request to the GCS in control, with a timeout specified in "Request timeout (sec)", in this case 7 seconds. This timeout will be sent as a parameter in the command, to syncronize progress indicators in both GCS. 
**We can not sent another request until that timeout expires**. This way, after clicking "send request", we will see the following:
![Screenshot from 2025-02-08 16-28-14](https://github.com/user-attachments/assets/78136445-1d1d-4555-9a1d-03f7d644232c)

And in the current GCS in control, a popup will appear indicating the request, with a syncronized countdown:
![Screenshot from 2025-02-08 16-29-32](https://github.com/user-attachments/assets/40a05778-dc8b-46d9-9bf8-965b10b6e725)
Clicking "Ignore" will just discard the popup, with no consequences, and if we click "allow takeover" the requestor GCS will be able to adquire control directly, as in the previous section  "Control request procedure when takeover is allowed". However, we will see the following popup appear:

![Screenshot from 2025-02-08 16-32-13](https://github.com/user-attachments/assets/b66e27bc-fb49-48f5-adf4-f7fd3420113e)
**We have a timeout of 10 seconds after clicking "allow takeover" above**. If after those 10 seconds the requestor GCS didn't take control, automatic takeover will be set to disabled again. **This will happen regardless of discarding ( clicking Ignore ) on the panel above**. This is done as a security measure so operators don't forget they accepted a takeover that was never fully completed.

#### Aditional comments
A couple of new command line arguments were added, in order to make testing easier:
- Argument to specify system id con command line when launching the app, useful for sending 2 instances at the same time with different sysid regardless of offline settings
```
--system-id:254
```
- Argument to be able to run multiple instances at the same time, by default only one instance of QGC can run
```
--allow-multiple
```
 Using both arguments it is handier to test this in SITL

#### Testing status
This has been tested in SITL using the following Ardupilot branch https://github.com/ArduPilot/ardupilot/pull/29252

Here is a short video showing this testing:

https://github.com/user-attachments/assets/da300e0f-3f33-4e58-920b-b12d543cd523

For awareness @hamishwillee @rmackay9. Also @julianoes might want to take a look at it too, for Px4 support.

Thanks to https://harrisaerial.com and https://www.lincesystems.com for sponsoring this PR, and to @hamishwillee for his wise and valuable support, experience and patience during all the iterations that preceded this work.

HTRamsey commented:
`--allow-multiple` is only for testing right? There's probably a lot of potential issues with running multiple instances but I guess it doesn't matter that much if it's just meant for testing. https://github.com/itay-grudev/SingleApplication has a neat concept of primary/secondary instances that could be used to better support multiple instances in the future.

Davidsastresas commented:
@HTRamsey Yes, it is only meant for testing. I really can not imagine a situation on real world usage where you would want to have 2 GCS instances.

As it is a command line argument, I doubt users will boot QGC with this option without noticing in a production GCS.

I would really like to leave that option, I don't need multiple QGC instances often, but every now and then it helps to have during development, I think it is handy to have.

DonLakeFlyer commented:
I haven't look in detail yet, but a few things:
* development.xml things should only be supported in debug builds. So the addition of the indicator to the toolbar should be ifdef'ed.
* What's you thinking around `dim=false`? Not so thrilled about some dropdowns dim and some not. If it makes sense then I think it would make sense for all of them.
* Similar to development.xml I would make `--allow-multiple` debug build only.
* Can you update the docs or this as well (explaining debug only for now)

Davidsastresas commented:
@DonLakeFlyer 

- About dimming or not, first time we used that was for gimbal indicator, it is like that on Stable. I get your logic of doing all of them or none of them. If I had to choose I would choose to not dim. Sometimes it could be interesting for an operator to have one of those extended panels to visualize some extended telemetry, or in the case of gimbal control to have the controls handy. 

My logic to not dim this one, together with gimbal control, is that users might want to eventually leave the panel open momentarily during some part of the operation, and dim is inconvenient on such scenarios.

- About docs, @hamishwillee and I will finish this https://github.com/mavlink/mavlink-devguide/pull/556 when we get more testing from users hopefully in the incoming weeks, is that fine?

I pushed a couple off commits addressing the QT_DEBUG matter for controlIndicator.qml and --allow-multiple argument.

DonLakeFlyer commented:
> * About dimming or not...

Reasonable. I didn't even know not dimming was an option. Can you make them all not dim by changing main popup in MainRootWindow and then remove the option of dim or not. I think that's better overall.

> * About docs

Make sense

Davidsastresas commented:
@DonLakeFlyer I applied your latest comments, thanks for the review!

Davidsastresas commented on src/Vehicle/Vehicle.cc:
```diff
@@ -3987,6 +4001,124 @@ void Vehicle::_handleMessageInterval(const mavlink_message_t& message)
+    if (commandLong.target_system != MAVLinkProtocol::instance()->getSystemId()) {
+        return;
+    }
+    if (commandLong.command == MAV_CMD(32100)) { // MAV_CMD_REQUEST_OPERATOR_CONTROL
```
For some reason it would not pick it otherwise. I suspect in the new mavlink implementation for some reason we are not getting the ids on development.xml.

I spent some time tracing it but couldn't find the cause and I needed to move on.

If this works nicely we will move it to common soon probably, that's why I really didn't thought much about it. Makes sense?

DonLakeFlyer replied:
This pull missing a mavlink version update? Hence development.xml is using an old version which doesn't include these new commands. You should just need up update the mavlink commit.

You can see it here:
![Screenshot 2025-02-26 at 9 48 01 AM](https://github.com/user-attachments/assets/7866b49a-bbaf-4e03-bcec-e29e1a3d54f3)

Davidsastresas commented on src/Vehicle/Vehicle.cc:
```diff
@@ -3987,6 +4001,124 @@ void Vehicle::_handleMessageInterval(const mavlink_message_t& message)
+        safeRequestTimeoutSecs = SettingsManager::instance()->flyViewSettings()->requestControlTimeout()->cookedDefaultValue().toInt();
+    }
+    sendMavCommand(_defaultComponentId,
+                   MAV_CMD(32100),          // MAV_CMD_REQUEST_OPERATOR_CONTROL
```
same https://github.com/mavlink/qgroundcontrol/pull/12410#discussion_r1971619268

Davidsastresas commented on src/Vehicle/Vehicle.cc:
```diff
@@ -3987,6 +4001,124 @@ void Vehicle::_handleMessageInterval(const mavlink_message_t& message)
+    }
+    sendMavCommand(_defaultComponentId,
+                   MAV_CMD(32100),          // MAV_CMD_REQUEST_OPERATOR_CONTROL
+                   false,                   // Don't show errors, as per Mavlink control protocol Autopilot will report result failed prior to forwarding the request to the GCS in control.
```
We will never know if it is that of something went wrong on the request.

I guess such situations could be dealt with in the future if we ever iterate this "sub-protocol". For awareness @hamishwillee

DonLakeFlyer replied:
No response to a mavlink command is different than getting a failure return. The problem is you are using the generic version of sendMavCommand. You should be using sendMavCommandWithHandler which allows you to create custom handling of errors. This allows you to distinguish between a failure response and no response at all with `MavCmdResultFailureNoResponseToCommand`. That way in the case of no response you can tell the user that the vehicle did not respond to the request.

Davidsastresas replied:
Fixed in https://github.com/mavlink/qgroundcontrol/pull/12410/commits/862d647a204f330ce676933e301dd6b90fb8fd65

Davidsastresas commented:
@DonLakeFlyer thanks for the review. I addressed most of your requests, but some others required clarification, so I left the corresponding threads opened.

About UI and style, sorry about it, this was thought for 4.4 initially. I personally don't think it is a huge deal, because this popup, together with the gimbal one, are not quite the same as the other top toolbar indicators. The rest of indicators usually report telemetry, so the logic of the new UI makes total sense, but I am not sure I would change this one.

I am open to it though, so we can come back to this in the future if that is fine, to not hold the PR because of it?

DonLakeFlyer commented:
> About UI and style

No need to update UI in this pull. Once this goes through I'll take a look at it later on to see if anything needs updating.

julianoes commented on src/main.cc:
```diff
@@ -201,6 +203,18 @@ int main(int argc, char *argv[])
+    if (hasSystemId) {
+        bool ok;
+        int systemId = systemIdStr.toInt(&ok);
+        if (ok && systemId >= 0 && systemId <= 255) {  // MAVLink system IDs are 8-bit
+            qDebug() << "Setting MAVLink System ID to:" << systemId;
+            MAVLinkProtocol::instance()->setSystemId(systemId);
+        } else {
+            qDebug() << "Not setting MAVLink System ID. It must be between 0 and 255. Invalid system ID value:" << systemIdStr;
```
I think this should be between 1 and 255. 0 is not a valid system ID in my understand as 0 means "all systems".

Davidsastresas replied:
You are right, thanks for spotting it! Fixed in https://github.com/mavlink/qgroundcontrol/pull/12410/commits/a7f460c700b35c86bae6febd9667f144f742e37f

julianoes reviewed (commented):
Thanks for the screenshare, it looks great!

One thing that I find a bit confusing is that you have to do two actions: first request, and then secondly confirm. Is that really required?

I would have expected to get control immediately if I requested it and it gets approved.

Davidsastresas commented:
> Thanks for the screenshare, it looks great!
> 
> One thing that I find a bit confusing is that you have to do two actions: first request, and then secondly confirm. Is that really required?
> 
> I would have expected to get control immediately if I requested it and it gets approved.

Hi Julian, thanks for the review. I am not sure what exactly you mean, but:

- If takeover is allowed, when you take control it is taken directly, just with one click.
- If takeover is not allowed, you send a control request. This control request needs to be accepted by the GCS in control, which will allow takeover, but only for 10 seconds. If control is not taken on those 10 seconds, takeover will be set to not allowed again, to avoid the operator forgetting about it and leaving the system in "allow takeover" state without having set it that way explicitely.

I imagine you were referring to the second scenario. This is something @hamishwillee  and I discussed back at the time, and we ended up with that procedure.

I think the reasoning was that the "requesting GCS" does not know exactly when the GCS in control will grant control, so it could grant control when the "requesting GCS" is not expecting it. So this procedure of first allowing takeover, so "requesting GCS" can then get control instantly with a click made more sense.

Davidsastresas commented:
@DonLakeFlyer I spared some time today to look at your pending requests.

### About mavlink and magic numbers
I updated mavlink to the latest version, tried to rebuild, but it still fails if I use MAV_CMD_REQUEST_OPERATOR_CONTROL instead of the magic number MAV_CMD(32100).

I used to be familiarized with how mavlink is integrated in QGC, but it has changed so much, and each time I look at it has changed again, so I am sorry but I don't have the time to troubleshoot why it isn't picking the commands in development.xml.

### About handlers https://github.com/mavlink/qgroundcontrol/pull/12410#discussion_r1972079390
Great point, and I agree that is the neat way of doing it. However I don't think I will have time until late April to spend time on this again. Do you think that should prevent the PR getting in? It is fine if you do, I will fix it in April.

Thank you very much for your review.

Davidsastresas commented:
Another solution would be to leave this PR as draft, or close it, and do a PR against Stable QGC.

DonLakeFlyer commented:
> do a PR against Stable QGC.

That doesn't make sense since development.xml stuff is debug build only. Also 5.0 will be out in a couple months.

DonLakeFlyer commented:
> However I don't think I will have time until late April to spend time on this again. Do you think that should prevent the PR getting in?

Create an Issue for this, assign it to yourself. And then I'm fine with it going in like this.

Davidsastresas commented:
@DonLakeFlyer good point about stable, makes sense. I created the issue https://github.com/mavlink/qgroundcontrol/issues/12552

Thanks!

DonLakeFlyer commented:
I'll take a stab at the mavlink version upgrade...

DonLakeFlyer commented:
> I updated mavlink to the latest version, tried to rebuild, but it still fails if I use MAV_CMD_REQUEST_OPERATOR_CONTROL instead of the magic number MAV_CMD(32100).

It's pretty easy, it's just here. Not sure if you need a clean full build if you change that:
https://github.com/mavlink/qgroundcontrol/blob/master/cmake/CustomOptions.cmake#L37

DonLakeFlyer commented:
And then you need to `ifdef` code which references development.xml stuff to be debug only since development.xml is debug only. I bumped the mavlink tag, updated code to use real command things from headers and build a debug and it worked fine for me.

DonLakeFlyer commented:
> No wait, it was working now it's not.. still looking...

It's broken upstream. I'll fix that, and bump the mavlink in that pull.

DonLakeFlyer commented:
So once my fix goes through. You'll need to rebase. Then make sure to `#ifdef QGC_DAILY_BUILD` the important parts which light this thing up on debug builds only. Should mostly be the command handlers I think.

Davidsastresas commented:
Please note I did a mavlink update right as you described earlier today, right before https://github.com/mavlink/qgroundcontrol/pull/12410#issuecomment-2718283446 and it didn´t work for me.

About the QGC_DAILY_BUILD part, sure! I will be a couple of weeks away so maybe I can not service it on the next few days, but I will be back on April.

Thanks Don!

DonLakeFlyer commented:
@Davidsastresas Given the way the current mavlink headers work there doesn't seem to be any way to get access to a new MAV_CMD from development dialect. I'm in discussion with mavlink folks about it. So magic number needs to remain. Also we are in the process of wrapping up the 5.0 release. So if you want this in there it will need to be wrapped up soon.

Davidsastresas commented:
@DonLakeFlyer thanks for letting me know. 

My understanding is that Mavlink folks are open to moving this to common once it is working and tested, but for that we need to first deliver support for it in QGC and the autopilots, so hopefully we can soon get rid of the magic numbers.

I will take a look at all of this again on the next couple of weeks.

Thank you!

hamishwillee commented:
@DonLakeFlyer Can you take the current https://github.com/mavlink/c_library_v2 and see if it fixes the missing definitions.

There is a flaw in the build toolchain which means that for all.xml it will include either development.xml or ardupilotmega.xml command definitions, but not both (see https://github.com/ArduPilot/pymavlink/pull/544#discussion_r2069976980). To fix this I have added a dummy command in all.xml, which forces all commands in both files to be merged. This is not a great fix, but it should unblock you.

Please ping me if it works.

Davidsastresas commented:
@hamishwillee I tested your suggestion and it works, thanks! I added it to the current PR https://github.com/mavlink/qgroundcontrol/pull/12410/commits/43ca6d39e8d3b957d0a535ce6864ae87e10ef2df. As a consequence, there is no need to use magic numbers anymore @DonLakeFlyer, so I updated it here https://github.com/mavlink/qgroundcontrol/pull/12410/commits/fc1520439b6b7037a1db32898c7b068db0678a47.

I also left the PR updated to current master. Thanks!

Davidsastresas commented:
@DonLakeFlyer I also addressed your concern in https://github.com/mavlink/qgroundcontrol/pull/12410#discussion_r1960439485 in commit https://github.com/mavlink/qgroundcontrol/pull/12410/commits/862d647a204f330ce676933e301dd6b90fb8fd65.

I think as it is now it should have fixed your latest concerns about this PR which were:
- Magic numbers because development.xml wasn't working as it should
- What happens if vehicle does not respond to operator request command?

I know we are a bit tight for the 5.0 release you wanted to make, so it is fine to leave this out, but even if we leave this out for the incoming release, It would be good to merge it soon after that. I've had to rebase this PR a few times already, and it's super time consuming at the current rate the codebase is changing.

Thanks!

DonLakeFlyer commented:
All good, thanks

DonLakeFlyer commented:
@hamishwillee So what's the state of the mavlink for this support? Is it still in development or has it moved to regular mavlink?

Davidsastresas commented:
@DonLakeFlyer Thank you, much appreciated!

hamishwillee commented:
> @hamishwillee So what's the state of the mavlink for this support? Is it still in development or has it moved to regular mavlink?

@DonLakeFlyer Apologies - with only 2 days a week on this community I sometimes miss messages.

The state is that this is in development.xml and not in any flight stack.
Further, there was a lot of late-churn on the requirements from ArduPilot deciding they want to support multi-GCS control as well as single-operator-in control.

I believe I have a way forward that should address concerns in https://github.com/mavlink/mavlink/pull/2313 but am still waiting for ArduPilot to properly engage. 

So the interface is likely to change. From a GCS perspective my proposal is a relatively small change, but it would be breaking,

Upshot, this must not go into a release. Its great that it is available for broad testing though.
